Modules/Devices/src/Devices.Infrastructure/PushNotifications/DirectPush/DirectPushService.cs

```diff
@@ -56,9 +57,15 @@ public async Task UpdateRegistration(IdentityAddress address, DeviceId deviceId,
             registration = new PnsRegistration(address, deviceId, handle, appId);
             pnsConnector.ValidateRegistration(registration);
 
-            await _pnsRegistrationRepository.Add(new PnsRegistration(address, deviceId, handle, appId), cancellationToken);
-
-            _logger.LogTrace("New device successfully registered.");
+            try
+            {
+                await _pnsRegistrationRepository.Add(new PnsRegistration(address, deviceId, handle, appId), cancellationToken);
+                _logger.LogTrace("New device successfully registered.");
+            }
+            catch (InfrastructureException exception)
```

Comment:
What if at some point we decide to throw an InfrastructureException for another reason? Then we would just log it here instead of passing it to the caller. So you should check for the specific error.
And can you please override the `Equals` method on the `InfrastructureError` and make it return true if the `Code` property is the same? (remember you can use ReSharper to generate the Equality members)